### DISCUSSION TITLE
what the difference between kfunc and kprobe

### DISCUSSION THREAD

**Discussion by ChinaXing** [bpftrace/bpftrace#1894] [Q&A] [answered]
what the difference between kfunc and kprobe

**Comment by fbs** [comment, marked as answer]
Kprobes are the classic kprobes.
- It's a probe somewhere in a kernel function, it can be at entry or any random instruction in it using the `k:f+0x10` syntax
- It doesn't have type info, all you get is `argX` which you have to cast to the appropriate type

kfuncs are the 'next gen' probes:
- they use btf for type info so types are known at compile time. Meaning you can write `args->pid` and bpftrace will figure out the  type and arg offset
- they *cannot* be attached at an offset, `kfunc:f+0x10` won't work
- they maintain a copy of the input args, so you can access the `args` in a `kretfunc`


Also see #1833, https://lwn.net/Articles/804112/
